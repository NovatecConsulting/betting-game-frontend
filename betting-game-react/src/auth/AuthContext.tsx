import React from "react";
import {
    User as OidcUser,
    UserManager,
    UserManagerSettings,
} from "oidc-client";

// For immutability
class User extends OidcUser {
    constructor(user: OidcUser) {
        super({
            id_token: user.id_token,
            access_token: user.access_token,
            expires_at: user.expires_at,
            profile: user.profile,
            refresh_token: user.refresh_token!,
            scope: user.scope,
            session_state: user.session_state!,
            state: user.state,
            token_type: user.token_type,
        });
    }
}

// TODO: error handling
/**
 * Manages the Authorization Code Flow with PKCE. Added such that underlying oidc library can be exchanged very easy
 * without being forced to change other parts of the app.
 */
class AuthorizationCodeFlowPKCEManager {
    private userManager: UserManager;
    private user: User | null;

    constructor(settings: UserManagerSettings) {
        this.userManager = new UserManager(settings);
        this.user = null;
    }

    public login = async () => {
        await this.userManager.clearStaleState();
        await this.userManager.signinRedirect();
    };

    public loginCallback = async () => {
        await this.userManager.signinRedirectCallback();
        const oidcUser = await this.userManager.getUser();
        this.user = oidcUser ? new User(oidcUser) : oidcUser;
    };

    public logout = async () => {
        // this causes an error at the moment, because there is a CORS bug at auth0
        // if a user is logged out the corresponding session and all refresh_tokens etc are revoked. on localhost this
        // has the effect, that users have to consent the access to their email on each login.
        // This is not happening on a real domain.
        // FIXME: add again
        // await this.userManager.revokeAccessToken();
        await this.userManager.signoutRedirect();
    };

    public updateToken = async (minValidity: number = 5) => {
        if (
            this.user &&
            this.user.refresh_token &&
            (this.user.expired || this.user.expires_in === minValidity)
        ) {
            try {
                await this.userManager.signinSilent();
                const oidcUser = await this.userManager.getUser();
                this.user = oidcUser ? new User(oidcUser) : oidcUser;
            } catch (e) {
                console.error(
                    "User context could not be refreshed. Logging out.",
                    e
                );
                await this.logout();
            }
        }
    };

    public getUser = () => this.user;
}

export let AuthorizationManager: AuthorizationCodeFlowPKCEManager;

const AuthContext: React.FC = ({ children }) => {
    // TODO: use config
    AuthorizationManager = new AuthorizationCodeFlowPKCEManager({
        authority: "https://betting-game.eu.auth0.com",
        client_id: "rqlyAphr2es3RUjYlPAEkf2VCldAR5wL",
        redirect_uri: "http://localhost:3000/login-callback",
        response_type: "code",
        response_mode: "query", // TODO: in the future hopefully "fragment" is supported
        scope: "openid profile email offline_access",
        post_logout_redirect_uri: "http://localhost:3000/",
        revokeAccessTokenOnSignout: true,
        metadata: {
            issuer: "https://betting-game.eu.auth0.com/",
            authorization_endpoint:
                "https://betting-game.eu.auth0.com/authorize",
            userinfo_endpoint: "https://betting-game.eu.auth0.com/userinfo",
            token_endpoint: "https://betting-game.eu.auth0.com/oauth/token",
            registration_endpoint:
                "https://betting-game.eu.auth0.com/oidc/register",
            revocation_endpoint:
                "https://betting-game.eu.auth0.com/oauth/revoke",
            end_session_endpoint:
                "https://betting-game.eu.auth0.com/v2/logout?client_id=rqlyAphr2es3RUjYlPAEkf2VCldAR5wL&returnTo=http://localhost:3000/",
            jwks_uri: "https://betting-game.eu.auth0.com/",
        },
    });
    return <>{children}</>;
};

export default AuthContext;
