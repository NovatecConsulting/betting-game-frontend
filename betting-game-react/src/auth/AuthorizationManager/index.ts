export { default as AuthorizationCodeFlowPKCEManager } from "./AuthorizationManager";
