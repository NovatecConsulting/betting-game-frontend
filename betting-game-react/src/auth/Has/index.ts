export { default as Has } from "./Has";
