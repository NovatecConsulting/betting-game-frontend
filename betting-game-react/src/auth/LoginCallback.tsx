import React, { useState } from "react";
import { Redirect } from "react-router-dom";
import { AuthorizationManager } from ".";
import { ROUTE_INDEX } from "../routes";

const LoginCallback: React.FC = () => {
    const [loggedIn, setLoggedIn] = useState(false);
    const [error, setError] = useState<string | undefined>(undefined);
    React.useEffect(() => {
        const doLoginAndRedirect = async () => {
            try {
                await AuthorizationManager.loginCallback();
                setLoggedIn(true);
            } catch (e) {
                setError(e.message);
            }
        };
        doLoginAndRedirect();
    }, []);
    // TODO: use Loading Component
    return error ? (
        <div>{error}</div>
    ) : loggedIn ? (
        <Redirect to={ROUTE_INDEX} />
    ) : (
        <div>Loading...</div>
    );
};

export default LoginCallback;
