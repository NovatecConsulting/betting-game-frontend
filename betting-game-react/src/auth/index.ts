export { default as AuthContext, AuthorizationManager } from "./AuthContext";
export { default as LoginCallback } from "./LoginCallback";
export * from "./permissions";
