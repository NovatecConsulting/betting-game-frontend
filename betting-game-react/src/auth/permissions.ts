export const READ_TEAMS = "read:teams";
export const READ_MATCHDAYS = "read:matchdays";
