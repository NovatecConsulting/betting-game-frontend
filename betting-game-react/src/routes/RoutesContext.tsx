import React, { useContext } from "react";
import { ExtendedRouteProps } from ".";

export const RoutesContext = React.createContext<ExtendedRouteProps[]>([]);
export const useRoutes = () => useContext(RoutesContext);
