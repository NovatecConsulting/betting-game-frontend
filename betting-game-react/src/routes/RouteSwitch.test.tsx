import React from "react";
import { shallow } from "enzyme";
import { RouteSwitch, ExtendedRouteProps } from ".";
import { useRoutes } from "./RoutesContext";
import { useHistory } from "react-router-dom";
import { useKeycloak } from "@react-keycloak/web";
import { Permission } from "../auth";

jest.mock("./RoutesContext");
jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useHistory: jest.fn()
}));
jest.mock("@react-keycloak/web");

class TEST_PERMISSION extends Permission {
    public static ACTION = new TEST_PERMISSION("ACTION");
}

describe(`${RouteSwitch.name} component`, () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    it("should render and match snapshot, given no routes.", () => {
        const history = { listen: jest.fn() };
        const keycloak = { updateToken: jest.fn() };
        (useRoutes as jest.Mock).mockReturnValueOnce([]);
        (useHistory as jest.Mock).mockReturnValueOnce(history);
        (useKeycloak as jest.Mock).mockReturnValueOnce({ keycloak });

        const wrapper = shallow(<RouteSwitch />);

        expect(history.listen).toHaveBeenCalled();
        expect(wrapper).toMatchSnapshot();
    });

    it("should render and match snapshot, given one route.", () => {
        const routes: ExtendedRouteProps[] = [
            {
                isNavItem: false,
                path: "/test",
                title: "Test",
                component: () => <></>,
                exact: true
            }
        ];
        const history = { listen: jest.fn() };
        const keycloak = { updateToken: jest.fn() };
        (useHistory as jest.Mock).mockReturnValueOnce(history);
        (useKeycloak as jest.Mock).mockReturnValueOnce({ keycloak });
        (useRoutes as jest.Mock).mockReturnValueOnce(routes);

        const wrapper = shallow(<RouteSwitch />);

        expect(wrapper).toMatchSnapshot();
    });

    it("should render and match snapshot, given two routes.", () => {
        const routes: ExtendedRouteProps[] = [
            {
                isNavItem: false,
                path: "/test",
                title: "Test",
                component: () => <></>,
                exact: true
            },
            {
                isNavItem: true,
                path: "/foo",
                title: "Foo",
                component: () => <></>,
                exact: true,
                hasPermission: TEST_PERMISSION.ACTION
            }
        ];
        const history = { listen: jest.fn() };
        const keycloak = { updateToken: jest.fn() };
        (useHistory as jest.Mock).mockReturnValueOnce(history);
        (useKeycloak as jest.Mock).mockReturnValueOnce({ keycloak });
        (useRoutes as jest.Mock).mockReturnValueOnce(routes);

        const wrapper = shallow(<RouteSwitch />);

        expect(wrapper).toMatchSnapshot();
    });
});
