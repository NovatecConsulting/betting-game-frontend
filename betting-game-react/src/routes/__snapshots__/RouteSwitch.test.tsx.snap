// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteSwitch component should render and match snapshot, given no routes. 1`] = `<Switch />`;

exports[`RouteSwitch component should render and match snapshot, given one route. 1`] = `
<Switch>
  <Route
    component={[Function]}
    exact={true}
    key="0"
    path="/test"
  />
</Switch>
`;

exports[`RouteSwitch component should render and match snapshot, given two routes. 1`] = `
<Switch>
  <Route
    component={[Function]}
    exact={true}
    key="0"
    path="/test"
  />
  <Route
    component={[Function]}
    exact={true}
    key="1"
    path="/test2"
  />
</Switch>
`;
